File: src/python/signalbird/_http.py
"""Ortak taşıma katmanı.

Bağımlılığı yoktur: standart kütüphanenin ``urllib``'i kullanılır. Bir SDK'nın
``requests`` sürümü dayatması, müşterinin bağımlılık çözümlemesini kilitleyen
en sık sebeptir — bir log/gönderim kütüphanesi bunu yapmamalı.

Zarf ve hata kodları diğer dillerle BİREBİR aynıdır (docs/CONTRACT.md § 8.2).
"""

from __future__ import annotations

import json
import socket
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Mapping, Optional

DEFAULT_BASE_URL = "https://signalbird.io/api"


class SignalbirdError(Exception):
    """``throw_on_error=True`` iken fırlatılır."""

    def __init__(self, message: str, status: int = 0, code: Optional[str] = None, body: Any = None):
        super().__init__(message)
        self.status = status
        self.code = code
        self.body = body


class Result(dict):
    """``{ok, status, data, code, message}`` zarfı.

    ``dict``'ten türer ki hem ``result["data"]`` hem ``result.ok`` çalışsın;
    diğer dillerdeki alan adlarıyla birebir kalır.
    """

    @property
    def ok(self) -> bool:
        return bool(self.get("ok"))

    @property
    def status(self) -> int:
        return int(self.get("status", 0))

    @property
    def data(self) -> Any:
        return self.get("data")

    @property
    def code(self) -> Optional[str]:
        return self.get("code")

    @property
    def message(self) -> Optional[str]:
        return self.get("message")


def build_query(query: Optional[Mapping[str, Any]]) -> str:
    """``None`` alanlar atlanır; diziler ``key[]=`` biçiminde gider."""
    if not query:
        return ""

    pairs = []

    for key, value in query.items():
        if value is None:
            continue

        if isinstance(value, (list, tuple, set)):
            for item in value:
                pairs.append((f"{key}[]", _stringify(item)))
        else:
            pairs.append((key, _stringify(value)))

    return "?" + urllib.parse.urlencode(pairs) if pairs else ""


def _stringify(value: Any) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"

    return str(value)


class Transport:
    """Anahtarlı istemcilerin ortak HTTP katmanı."""

    def __init__(self, api_key: str, base_url: str, timeout: float, throw_on_error: bool, debug: bool):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.throw_on_error = throw_on_error
        self.debug = debug

    def request(
        self,
        method: str,
        path: str,
        body: Optional[Mapping[str, Any]] = None,
        query: Optional[Mapping[str, Any]] = None,
        auth_header: str = "Authorization",
        auth_prefix: str = "Bearer ",
    ) -> Result:
        url = self.base_url + path + build_query(query)
        payload = None
        headers = {
            "Accept": "application/json",
            auth_header: f"{auth_prefix}{self.api_key}",
        }

        if body is not None:
            payload = json.dumps(_drop_none(body), ensure_ascii=False).encode("utf-8")
            headers["Content-Type"] = "application/json"

        request = urllib.request.Request(url, data=payload, headers=headers, method=method)

        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                return self._success(response.status, _decode(response.read()))
        except urllib.error.HTTPError as error:
            data = _decode(error.read())
            status = int(error.code)

            # API `{message, code}` döner; Laravel doğrulama hatası `{message,
            # errors}` döner (kodsuz) — onu VALIDATION_ERROR sayarız.
            code = None
            message = None

            if isinstance(data, dict):
                code = data.get("code") if isinstance(data.get("code"), str) else None
                message = data.get("message") if isinstance(data.get("message"), str) else None

            if not code:
                code = (
                    "VALIDATION_ERROR"
                    if status == 422
                    else "API_KEY_INVALID"
                    if status == 401
                    else f"HTTP_{status}"
                )

            return self._fail(status, code, message or f"HTTP {status}", data)
        except socket.timeout:
            return self._fail(0, "TIMEOUT", "request timed out", None)
        except urllib.error.URLError as error:
            reason = getattr(error, "reason", error)

            if isinstance(reason, socket.timeout):
                return self._fail(0, "TIMEOUT", "request timed out", None)

            return self._fail(0, "NETWORK_ERROR", str(reason), None)

    def _success(self, status: int, data: Any) -> Result:
        return Result(ok=True, status=status, data=data, code=None, message=None)

    def _fail(self, status: int, code: str, message: str, data: Any) -> Result:
        if self.throw_on_error:
            raise SignalbirdError(f"Signalbird: {code} — {message}", status, code, data)

        if self.debug:
            print(f"[signalbird] {code} (HTTP {status}): {message}")

        return Result(ok=False, status=status, data=data, code=code, message=message)


def _decode(raw: bytes) -> Any:
    if not raw:
        return None

    text = raw.decode("utf-8", errors="replace")

    try:
        return json.loads(text)
    except ValueError:
        return text


def _drop_none(body: Mapping[str, Any]) -> dict:
    """``None`` alanlar gövdeden atılır: "gönderilmedi" ile "null yapıldı" aynı şey değil."""
    return {key: value for key, value in body.items() if value is not None}


def seg(value: Any) -> str:
    """Yol parçası — kimlikler URL'e gömülmeden önce kodlanır."""
    return urllib.parse.quote(str(value), safe="")

File: src/go/signalbird/http.go
// Package signalbird — Signalbird SDK'sının Go istemcisi.
//
// Üç sunucu yüzeyi vardır ve anahtarları farklıdır:
//
//	Client     → Telsiz (log yazma),  sbr_live_…
//	Messaging  → Gönderim,            sb_…
//	Management → Yönetim,             sb_… + radio|chat|apps scope'ları
//
// Bağımlılığı yoktur: yalnız standart kütüphane. Retry yoktur — aynı iletiyi
// iki kez göndermek hiç göndermemekten pahalıdır, yeniden deneme kararı
// çağıranındır.
//
// Sözleşme: docs/CONTRACT.md
package signalbird

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// DefaultBaseURL — kendi kurulumu olan müşteri BaseURL ile değiştirebilir.
const DefaultBaseURL = "https://live.signalbird.io/api"

// Result, her metodun döndüğü zarftır. Başarısızlık hata değil, veridir:
// log göndermek ya da bir kaydı okumak uygulamanın ASIL işi değildir.
type Result struct {
	OK      bool            `json:"ok"`
	Status  int             `json:"status"`
	Data    json.RawMessage `json:"data,omitempty"`
	Code    string          `json:"code,omitempty"`
	Message string          `json:"message,omitempty"`
}

// Into, Data alanını verilen yapıya çözer.
//
//	var project struct{ Project struct{ ID int } }
//	res.Into(&project)
func (r Result) Into(target any) error {
	if len(r.Data) == 0 {
		return nil
	}

	return json.Unmarshal(r.Data, target)
}

// Error, ThrowOnError açıkken döner.
type Error struct {
	Code    string
	Status  int
	Message string
	Body    json.RawMessage
}

func (e *Error) Error() string {
	return fmt.Sprintf("signalbird: %s (HTTP %d): %s", e.Code, e.Status, e.Message)
}

// ErrWrongKeyType — istemciye yanlış aileden anahtar verildi.
var ErrWrongKeyType = errors.New("signalbird: yanlış anahtar türü")

// ErrNoKey — anahtar boş.
var ErrNoKey = errors.New("signalbird: anahtar zorunlu")

type transport struct {
	apiKey       string
	baseURL      string
	authHeader   string
	authPrefix   string
	client       *http.Client
	throwOnError bool
	debug        bool
}

func newTransport(apiKey, baseURL string, timeout time.Duration, throwOnError, debug bool) *transport {
	if baseURL == "" {
		baseURL = DefaultBaseURL
	}

	return &transport{
		apiKey:       apiKey,
		baseURL:      strings.TrimRight(baseURL, "/"),
		authHeader:   "Authorization",
		authPrefix:   "Bearer ",
		client:       &http.Client{Timeout: timeout},
		throwOnError: throwOnError,
		debug:        debug,
	}
}

// Query — sorgu dizesi. nil değerler atlanır, dilimler key[]= biçiminde gider.
type Query map[string]any

func (t *transport) request(ctx context.Context, method, path string, body any, query Query) (Result, error) {
	var reader io.Reader

	if body != nil {
		encoded, err := json.Marshal(body)
		if err != nil {
			return t.fail(0, "ENCODE_ERROR", err.Error(), nil)
		}

		reader = bytes.NewReader(encoded)
	}

	if ctx == nil {
		ctx = context.Background()
	}

	request, err := http.NewRequestWithContext(ctx, method, t.baseURL+path+buildQuery(query), reader)
	if err != nil {
		return t.fail(0, "NETWORK_ERROR", err.Error(), nil)
	}

	request.Header.Set("Accept", "application/json")
	request.Header.Set(t.authHeader, t.authPrefix+t.apiKey)

	if body != nil {
		request.Header.Set("Content-Type", "application/json")
	}

	response, err := t.client.Do(request)
	if err != nil {
		code := "NETWORK_ERROR"
		if errors.Is(err, context.DeadlineExceeded) || strings.Contains(err.Error(), "Client.Timeout") {
			code = "TIMEOUT"
		}

		return t.fail(0, code, err.Error(), nil)
	}
	defer response.Body.Close()

	raw, _ := io.ReadAll(response.Body)

	if response.StatusCode >= 200 && response.StatusCode < 300 {
		return Result{OK: true, Status: response.StatusCode, Data: json.RawMessage(raw)}, nil
	}

	// API {message, code} döner; Laravel doğrulama hatası {message, errors}
	// döner (kodsuz) — onu VALIDATION_ERROR sayarız.
	var parsed struct {
		Code    string `json:"code"`
		Message string `json:"message"`
	}
	_ = json.Unmarshal(raw, &parsed)

	code := parsed.Code
	if code == "" {
		switch response.StatusCode {
		case 422:
			code = "VALIDATION_ERROR"
		case 401:
			code = "API_KEY_INVALID"
		default:
			code = fmt.Sprintf("HTTP_%d", response.StatusCode)
		}
	}

	message := parsed.Message
	if message == "" {
		message = fmt.Sprintf("HTTP %d", response.StatusCode)
	}

	return t.fail(response.StatusCode, code, message, raw)
}

func (t *transport) fail(status int, code, message string, body []byte) (Result, error) {
	result := Result{OK: false, Status: status, Code: code, Message: message, Data: json.RawMessage(body)}

	if t.throwOnError {
		return result, &Error{Code: code, Status: status, Message: message, Body: result.Data}
	}

	if t.debug {
		fmt.Printf("[signalbird] %s (HTTP %d): %s\n", code, status, message)
	}

	return result, nil
}

func buildQuery(query Query) string {
	if len(query) == 0 {
		return ""
	}

	values := url.Values{}

	for key, value := range query {
		if value == nil {
			continue
		}

		switch typed := value.(type) {
		case []string:
			for _, item := range typed {
				values.Add(key+"[]", item)
			}
		case []int:
			for _, item := range typed {
				values.Add(key+"[]", fmt.Sprint(item))
			}
		case bool:
			if typed {
				values.Set(key, "true")
			} else {
				values.Set(key, "false")
			}
		default:
			values.Set(key, fmt.Sprint(typed))
		}
	}

	if len(values) == 0 {
		return ""
	}

	return "?" + values.Encode()
}

// seg — kimlikler URL'e gömülmeden önce kodlanır.
func seg(value any) string {
	return url.PathEscape(fmt.Sprint(value))
}
